### ISSUE TITLE
Port twisted.python.versions to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5917]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5917|
|Type|enhancement|
|Created|2012-08-24 19:17:47Z|
|Branch|https://github.com/twisted/twisted/tree/versions-python3-5917-3|

`twisted/__init__.py` would like this.  Also, [ticket:5916 `twisted.python.deprecate`] depends on `twisted.python.versions`.

<details><summary>Searchable metadata</summary>

```
trac-id__5917 5917
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_versions_python3_5917_3 branches/versions-python3-5917-3
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1345835867000000 1345835867000000
changetime__1346241741000000 1346241741000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [35386]) Branching to 'versions-python3-5917'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [35406]) Branching to 'versions-python3-5917-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [35409]) Switch to byte literals in a bunch of places.  Do a little messing about with encodings.  Close files instead of relying on the gc to do it.  Add twisted.python.versions and twisted.python.test.test_versions to the ported list.

refs [#5917](https://github.com/twisted/twisted/issues/5917)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [35417]) Branching to 'versions-python3-5917-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @itamarst|
|-|-|

Some stuff done, [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/versions-python3-5917-3) (may they be favorable).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst set owner to @exarkun|
|-|-|

Sorry for the late review.

  1. Close [#5785](https://github.com/twisted/twisted/issues/5785) as a duplicate.
  1. Given [#5886](https://github.com/twisted/twisted/issues/5886) is pretty trivial, maybe add it to this branch.
  1. Needs news file.
  1. Having the SVN revision as bytes means you get funny looking str(), e.g. 

```
>>> str(v)
"[twisted, version 1.0.0+rb'35423']"
```
     So you probably want to `nativeString` it at some point in the code.

Fix those and merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [35426]) re-enable twisted.__version__

refs [#5917](https://github.com/twisted/twisted/issues/5917)
refs [#5886](https://github.com/twisted/twisted/issues/5886)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [35427]) Add a test for getVersionString result including an svn revision number, and make it pass on Python 3.x

refs [#5917](https://github.com/twisted/twisted/issues/5917)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [35428]) Fix pyflakes warning

refs [#5917](https://github.com/twisted/twisted/issues/5917)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [35431]) Merge versions-python3-5917-3

Author: exarkun
Reviewer: itamar
Fixes: [#5917](https://github.com/twisted/twisted/issues/5917)
Fixes: [#5785](https://github.com/twisted/twisted/issues/5785)
Fixes: [#5886](https://github.com/twisted/twisted/issues/5886)

Port `twisted.python.versions` to Python 3.  Also add a bit of missing test coverage
and restore the definition of `twisted.__version__` on Python 3.
